contracts/TimeManager.sol

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.13;
+
+import { SECONDS_PER_YEAR } from "./lib/constants.sol";
+
+abstract contract TimeManager {
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    uint256 public immutable blocksPerYear;
+
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    bool public immutable isTimeBased;
+
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    function() view returns (uint256) internal immutable _getCurrentSlot;
+
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    function() view returns (uint256) internal immutable _getSlotsInOneYear;
+
+    /// @notice Thrown on invaid arguments
+    error InvalidBlocksPerYear();
+
+    /**
+     * @param timeBased_ A boolean indicating whether the contract is based on time or block.
+     * @param blocksPerYear_ The number of blocks per year
+     * @custom:error InvalidBlocksPerYear is thrown if blocksPerYear entered is zero
+     * @custom:oz-upgrades-unsafe-allow constructor
+     */
+    constructor(bool timeBased_, uint256 blocksPerYear_) {
+        if (blocksPerYear_ == 0) {
```

Comment:
Should we force it to be exactly `SECONDS_PER_YEAR` for time based?

```suggestion
        if (blocksPerYear_ == 0 || (timeBased_ && blocksPerYear != SECONDS_PER_YEAR)) {
```

Reply:
https://github.com/VenusProtocol/isolated-pools/pull/324#discussion_r1411724475

Reply:
https://github.com/VenusProtocol/isolated-pools/commit/6b98b095e5e6b3322b6d0c825ba273612e49948a